### PR TITLE
Add optional mount args. Change the args for Azure disk mount.

### PR DESCRIPTION
**Pull request by brendandburns**
@rootfs @khenidak @jdumars

Related to https://github.com/kubernetes/kubernetes/issues/35180

### CODE REVIEW

File: pkg/util/mount/mount_linux.go [hunk 1]
```diff
@@ -367,7 +367,11 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
 			}
 
 			if fstype == "ext4" || fstype == "ext3" {
-				args = []string{"-F", source}
+				if len(mountArgs) == 0 {
+					args = []string{"-F", source}
+				} else {
+					args = append(mountArgs, source)
```
Comment (rootfs) [inline pkg/util/mount/mount_linux.go]: append to `options`? otherwise we need to have `-o` in mount command to proceed the mount options.

Comment (rootfs) [inline pkg/util/mount/mount_linux.go]: @brendandburns can you give examples of mount args? I was thinking of mount options, but you may want to pass something else.

File: pkg/volume/azure_dd/attacher.go [hunk 2]
```diff
@@ -224,7 +224,7 @@ func (attacher *azureDiskAttacher) MountDevice(spec *volume.Spec, devicePath str
 	if notMnt {
 		diskMounter := &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()}
 		mountOptions := volume.MountOptionFromSpec(spec, options...)
-		err = diskMounter.FormatAndMount(devicePath, deviceMountPath, *volumeSource.FSType, mountOptions)
+		err = diskMounter.FormatAndMount(devicePath, deviceMountPath, *volumeSource.FSType, mountOptions, []string{"-E", "lazy_itable_init=0,lazy_journal_init=0", "-F"})
```
Comment (rootfs) [inline pkg/volume/azure_dd/attacher.go]: [This comment](https://github.com/kubernetes/kubernetes/issues/30752#issuecomment-261303142) suggests setting lazy_itable_init=1 (i.e. default)

File: pkg/util/mount/mount.go [hunk 3]
```diff
@@ -84,14 +84,14 @@ type SafeFormatAndMount struct {
 // read-only it will format it first then mount it. Otherwise, if the
 // disk is already formatted or it is being mounted as read-only, it
 // will be mounted without formatting.
-func (mounter *SafeFormatAndMount) FormatAndMount(source string, target string, fstype string, options []string) error {
+func (mounter *SafeFormatAndMount) FormatAndMount(source string, target string, fstype string, options []string, mountArgs []string) error {
```
Comment (rootfs) [inline pkg/util/mount/mount.go]: `formatArgs` looks more appropriate based on the context.

**Comment by marun** [comment]
Not entirely sure which sig should be primary on this, feel free to change.

**Comment by jdumars** [comment]
@marun SIG Azure is watching it, but SIG Storage definitely has a stake in it as well.

File: pkg/volume/azure_dd/attacher.go (hunk 2 above)
Comment (brendandburns) [inline pkg/volume/azure_dd/attacher.go]: @khenidak what do you think? This is copied from your PR...

Comment (rootfs) [inline pkg/volume/azure_dd/attacher.go]: For reference, I formatted a 100GB Premium SSD on an F instance, repeated 3 times and time was consistent:
```console
# time mkfs.ext4 -F -E lazy_itable_init=0,lazy_journal_init=0 /dev/sdf 2>&1 > /dev/null
mke2fs 1.42.9 (28-Dec-2013)

real    0m18.770s
user    0m0.172s
sys     0m1.207s

# time mkfs.ext4 -F  /dev/sdf 2>&1 > /dev/null
mke2fs 1.42.9 (28-Dec-2013)

real    0m1.840s
user    0m0.158s
sys     0m0.118s
```

**Comment by jdumars** [comment]
@khenidak thoughts on the above?

File: pkg/volume/azure_dd/attacher.go (hunk 2 above)
Comment (khenidak) [inline pkg/volume/azure_dd/attacher.go]: I think we should avoid lazy_itable_init=1 as disks might gets quickly detached (i.e pod crash, node crash) which can give us unknown behavior if itable init was interrupted.

**Comment by khenidak** [comment]
I have the following 3 comments/discussion points:
1. Adding mountArgs/formatArgs will break things on cross OS implementation. Plugin developers will have to create conditional statements for each OS they want to support their plugin on. I think it takes away the value of mount/format abstraction currently implemented. Note: I realize that Azure PD implementation currently have a hack that assumes only linux, but that is one plugin (not the entire set of plugins).  

2. mkfs parameters were actually enabled before (**not sure why it is gone on HEAD**) https://github.com/kubernetes/kubernetes/blob/13a2bc8afb6fd19a34ebabd33a52845161a1b63a/pkg/util/mount/mount_linux.go.

3. The problem i was trying to solve for was the sequence of events not the parameters of mkfs.ext4. The sequence of fsck->mount->if failed format hangs (well the process goes into "waiting for I/O" for 3 to 6+ mins before it comes back). This was not unacceptable. I was under the impression that this is an Azure specific issue. Hence the original hack we had in the new PD implementation (Which does not attempt to execute mount command until disk is formatted (if it was not formatted). This gave us consistent behavior of < 1min of the entire PD sequence (attach, wait, format, mount). Less if the disk was already formatted.  @rootfs  wanted this be pushed down the stack in mount_linux.go implementation. This PR does not do that.

**Comment by rootfs** [comment]
> mkfs parameters were actually enabled before (not sure why it is gone on HEAD) https://github.com/kubernetes/kubernetes/blob/13a2bc8afb6fd19a34ebabd33a52845161a1b63a/pkg/util/mount/mount_linux.go.

@khenidak #38865 removed lazy_itable_init to address slow mkfs.ext4 issue on Azure #30752.

**Comment by khenidak** [comment]
Thanks @rootfs for the pointers. If i understand correctly?  same mkfs.ext4 command + parameters inside containerized kubelet is slow, outside is fast. Yet some how we ended at enabling lazy_itables_init 

We have reports that with 1.6 (where lazy_itables_init is enabled) we still run on a 3+ mins format time per PD? @colemickens can you confirm this?

**Comment by rootfs** [comment]
> We have reports that with 1.6 (where lazy_itables_init is enabled) we still run on a 3+ mins format time per PD? @colemickens can you confirm this?

The root cause for 3+ mins format time is likely related to caching mode. It currently defaults to [None](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_dd.go#L135).

I did a A/B test with caching mode None and ReadWrite. Both tests provisioned 300GB Standard_LRS disk. With Caching mode None, the format time was 4 mins, while with Caching mode ReadWrite, format finished in 20 seconds.

I don't have any reference on impact of Caching mode, but I am open to flip the default to ReadWrite. 


@brendandburns @khenidak @colemickens

**Comment by codablock** [comment]
Just stumbled upon the discussion about lazy init here. The (non-)lazy init problem and the kubelet IO performance degradation (https://github.com/kubernetes/kubernetes/issues/30752) were/are actually two different problems. https://github.com/kubernetes/kubernetes/pull/38865 has a good description about the lazy init problem. The kubelet IO degradation has other reasons (which I don't know atm). Both are already bad alone, but combined they resulted in multiple hours formatting time for volume with multiple hundreds of GB size. With the lazy init PR, that was reduced to a few minutes (due to kubelet), which is still not so nice, but at least not so bad as it was before.

Also, the partition misalignment that was suggested as the reason behind the performance degradation turned out to be not the reason as there aren't even partitions in play for Azure.

I have the feeling (did not test this, so really just a feeling!), that the kubelet performance degradation problem is not only still there, but that it is also present on all cloud providers. I assume it was only noticed on Azure because the (non-)lazy init problem worsened it so much.

File: pkg/volume/azure_dd/attacher.go (hunk 2 above)
Comment (codablock) [inline pkg/volume/azure_dd/attacher.go]: The kernel internally keeps track of the lazy init progress and maybe even stores the progress on-disk. If the disk is later mounted on another node, the kernel will continue the lazy init.

**Comment by dchen1107** [comment]
@brendanburns I removed this from v1.7 since the original issue is not approved for the milestone. Also the change is relatively risk given we plan to cut the release next week. Can we merge this to the main branch, then cherrypick it to the patch release?

**Comment by gnufied** [comment]
@brendanburns Can we add a description about why `lazy_itable_init=0,lazy_journal_init=0` options are being brought back for Azure? As mentioned elsewhere I thought use of those options were one of the root cause of #30752

**Comment by jdumars** [comment]
@brendandburns is this PR still relevant?

**Pull request closed without merging**